sycl/test-e2e/Graph/RecordReplay/handlerless_enqueue_functions.cpp

```diff
@@ -0,0 +1,119 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// Extra run to check for leaks in Level Zero using UR_L0_LEAKS_DEBUG
+// RUN: %if level_zero %{%{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
+
+// Test recording of several handlerless SYCL queue APIs (memset, memcpy,
+// parallel_for/nd_launch, single_task) inside a single function call
+// while graph recording is active. All operations are USM-based and occur via
+// eventless queue free functions or eventful queue shortcuts to bypass handler
+// path. Recording is performed over a non-inlined function call.
+
+#include "../graph_common.hpp"
+#include <sycl/ext/oneapi/experimental/enqueue_functions.hpp>
+#include <sycl/properties/all_properties.hpp>
+
+// noinline is important as we have previously caught functional issues with
+// kernel argument capture only when the function being recorded is not inlined.
+__attribute__((noinline)) void
+recordHandlerLessOps(sycl::queue &Q, uint32_t *A, uint32_t *B, uint32_t *C,
+                     uint32_t *D, uint32_t *E, size_t N, unsigned char Pattern,
+                     uint32_t FillValue, bool InOrderQueue) {
+  size_t WorkGroupSize = 16;
+  sycl::nd_range<1> KernelRange{sycl::range<1>{N},
+                                sycl::range<1>{WorkGroupSize}};
+  auto DoubleKernelLambda = [=](sycl::nd_item<1> item) {
+    const size_t i = item.get_global_linear_id();
+    C[i] = B[i] * 2;
+  };
+  auto SingleTaskKernel = [=]() { C[0] = 999; };
+  // Test eventless free functions with in-order queue and eventful shortcuts
+  // with out-of-order queue.
+  if (InOrderQueue) {
+    exp_ext::memset(Q, A, Pattern, N * sizeof(uint32_t));
+    exp_ext::fill(Q, D, FillValue, N);
+    exp_ext::copy(Q, D, E, N);
+    exp_ext::memcpy(Q, B, A, N * sizeof(uint32_t));
+    exp_ext::nd_launch(Q, KernelRange, DoubleKernelLambda);
+    exp_ext::single_task(Q, SingleTaskKernel);
+  } else {
+    auto e1 = Q.memset(A, Pattern, N * sizeof(uint32_t));
+    auto e2 = Q.fill(D, FillValue, N);
+    Q.copy(D, E, N, e2);
+    auto e3 = Q.memcpy(B, A, N * sizeof(uint32_t), e1);
+    auto e4 = Q.parallel_for(KernelRange, e3, DoubleKernelLambda);
+    Q.single_task(e4, SingleTaskKernel);
```

Comment:
Looks like we have two independent streams of execution here. Would it make sense to reorder queue submissions to make this more obvious? or is there a specific reason why you picked that order ...

Reply:
There is no particular reason for the ordering. I reordered the submissions as suggested for clarity.